lib/vast/Conversion/Generic/LowerValueCategories.cpp

```diff
@@ -0,0 +1,559 @@
+        {
+            mlir::MLIRContext &mctx;
+
+            // Do we need data layout? Probably not as everything should be explicit
+            // type size wise.
+            value_category_type_converter(mcontext_t &mctx) : mctx(mctx) {
+                init();
+
+                addConversion(this->template convert_fn_type< core::FunctionType >());
+            }
+
+            using mixin_base = tc::mixins< value_category_type_converter >;
+            using mixin_base::convert_type_to_type;
+            using mixin_base::convert_type_to_types;
+
+            template< typename T, typename... Args >
+            auto make_aggregate_type(Args... args) {
+                return [=](mlir_type elem) { return T::get(elem.getContext(), elem, args...); };
+            }
+
+            auto convert_lvalue_type() {
+                return [&](hl::LValueType type) {
+                    return Maybe(type.getElementType())
+                        .and_then(convert_type_to_type())
+                        .unwrap()
+                        .and_then(make_aggregate_type< hl::PointerType >())
+                        .template take_wrapped< maybe_type_t >();
+                };
+            }
+
+            template< typename T >
+            auto pointerlike_as_memref() {
+                return [&](T type) {
+                    // It should never happen that we have nested lvalues?
+                    auto element_type = this->convert_type_to_type(type.getElementType());
+                    return hl::PointerType::get(&mctx, *element_type);
+                };
+            }
+
+            void init() {
+                addConversion([&](mlir_type t) { return t; });
+                addConversion(pointerlike_as_memref< hl::LValueType >());
+                addTargetMaterialization(
+                    [&](mlir::OpBuilder &builder, mlir::Type resultType,
+                        mlir::ValueRange inputs, mlir::Location loc) -> std::optional< Value > {
+                        if (inputs.size() != 1) {
+                            return std::nullopt;
+                        }
+
+                        return builder
+                            .create< mlir::UnrealizedConversionCastOp >(loc, resultType, inputs)
+                            .getResult(0);
+                    }
+                );
+                addSourceMaterialization(
+                    [&](mlir::OpBuilder &builder, mlir::Type resultType,
+                        mlir::ValueRange inputs, mlir::Location loc) -> std::optional< Value > {
+                        if (inputs.size() != 1) {
+                            return std::nullopt;
+                        }
+
+                        return builder
+                            .create< mlir::UnrealizedConversionCastOp >(loc, resultType, inputs)
+                            .getResult(0);
+                    }
+                );
+            }
+        };
+
+#define VAST_DEFINE_REWRITE \
```

Comment:
Add undef at end.

Reply:
It is inside a `.cpp` so it is not strictly needed but I will add it.